### DISCUSSION TITLE
How to deal with identifiers that are generated during run-time and isn't the result of a graph-service operation during testing (mocking)?

### DISCUSSION THREAD

**Discussion by joffinjoy** [Sunbird-inQuiry/Community#9] [Q&A] [answered]
The objective is to write a test-case that tests question-Set copy including copy of branching-Logic. 
The questionSet I am trying to copy is:
```
do_1234 (QS)
	-do_2222 (Section QS)
		do_5555 (Q, Default)
		do_7777 (Q, Default)
```
\
And my newRootNode() mocking function returns:
```
do_9876 (QS)
	-do_3333 (Section QS)
		do_5555 (Q, Default)
		do_7777 (Q, Default)
```
\
But the section QS doesn't have a real node in database right? So the execution is purely based on a "do" id that gets generated during the execution. And since it doesn't have an actual node, I can't return it during the mocking.

So I am getting something like:
```
do_9876 (QS)
	-do_11351130082151628813 (Section QS)
		do_5555 (Q, Default)
		do_7777 (Q, Default)
```
And as expected, the execution breaks with an exception: **Exception occurred - class :org.sunbird.common.exception.ResourceNotFoundException with message :No node found with id: do_11351130082151628813**

How can I deal with this in the test-case? I can't know what do_id will be generated during runtime. 
\
Note: _The questionSet has to be of this configuration because the test case is to specifically test the branchingLogic copy. So two questions within a section within an observation is the minimum questionset required._

**Comment by krgauraw** [comment, marked as answer]
@joffinjoy In this case, When 1st call to update hierarchy flow ends, it returns a response having older_do_ids: new_do_ids for all the nodes which are passed under nodesModified section of the request. Please use that response to figure out newly generated section identifier and use it in further flow (e.g: mock returns, etc). 
Hope this helps.
cc: @pallakartheekreddy

**Reply by joffinjoy** [reply]
@krgauraw Is it possible to get that response into my test-case so that I can generate new fake nodes for mocking? I am not aware of such a technique. Can you share a snippet or link with an example?

**Reply by krgauraw** [reply]
@joffinjoy you can  mock the cassandra writes and let update hierarchy flow executes , so you will have response. 
In case, if still it is not clear for you, please post your test case here because without having context of how exactly  test case is written, it's difficult to suggest the code snippet.
cc: @pallakartheekreddy @Krishnaj20

**Reply by joffinjoy** [reply]
@krgauraw 

```
it should "return success response for 'copyQuestionSet' (Branching Logic Copy)" in {
        implicit val oec: OntologyEngineContext = mock[OntologyEngineContext]
        val graphDB = mock[GraphService]
        (oec.graphService _).expects().returns(graphDB).anyNumberOfTimes()
        val nodes: util.List[Node] = getCategoryNode()
        (graphDB.getNodeByUniqueIds(_: String, _: SearchCriteria)).expects(*, *).returns(Future(nodes)).anyNumberOfTimes()
        (graphDB.getNodeByUniqueId(_: String, _: String, _: Boolean, _: Request)).expects(*, "do_1234", *, *).returns(Future(getRootNodeWithBL("do_1234", "do_2222", true, true))).anyNumberOfTimes()
        (graphDB.getNodeByUniqueId(_: String, _: String, _: Boolean, _: Request)).expects(*, "do_9876", *, *).returns(Future(getRootNodeWithBL("do_9876", "do_3333", false, false))).anyNumberOfTimes()
        (graphDB.getNodeByUniqueId(_: String, _: String, _: Boolean, _: Request)).expects(*, "do_9876.img", *, *).returns(Future(getRootNodeWithBL("do_9876", "do_3333", false, false))).anyNumberOfTimes()
        (graphDB.getNodeByUniqueId(_: String, _: String, _: Boolean, _: Request)).expects(*, "do_5555", *, *).returns(Future(getQuestionNodeBL("do_5555"))).anyNumberOfTimes()
        (graphDB.getNodeByUniqueId(_: String, _: String, _: Boolean, _: Request)).expects(*, "do_7777", *, *).returns(Future(getQuestionNodeBL("do_7777"))).anyNumberOfTimes()
        (graphDB.readExternalProps(_: Request, _: List[String])).expects(*, List("objectMetadata")).returns(Future(getSuccessfulResponse)).anyNumberOfTimes()
        (graphDB.readExternalProps(_: Request, _: List[String])).expects(*, List("instructions", "outcomeDeclaration")).returns(Future(getSuccessfulResponse)).anyNumberOfTimes()
        (graphDB.readExternalProps(_: Request, _: List[String])).expects(*, List("solutions", "body", "editorState", "interactions", "hints", "responseDeclaration", "media", "answer", "instructions")).returns(Future(getResourceNotFoundResponse)).anyNumberOfTimes()
        (graphDB.addNode(_: String, _: Node)).expects(*, *).returns(Future(getRootNodeWithBL("do_9876", "do_3333", false, false))).anyNumberOfTimes
        (graphDB.saveExternalProps(_: Request)).expects(*).returns(Future(getSuccessfulResponse)).anyNumberOfTimes
        (graphDB.updateExternalProps(_: Request)).expects(*).returns(Future(getSuccessfulResponse)).anyNumberOfTimes
        (graphDB.getNodeByUniqueId(_: String, _: String, _: Boolean, _: Request)).expects(*, *, *, *).returns(Future(getRootNodeWithBL("do_9876", "do_3333", true, true))).anyNumberOfTimes()
        inSequence {
            (graphDB.upsertNode(_: String, _: Node, _: Request)).expects(*, *, *).returns(Future(getUpsertNodeBLWithoutBL))
            (graphDB.upsertNode(_: String, _: Node, _: Request)).expects(*, *, *).returns(Future(getUpsertNodeBLWithBL)).anyNumberOfTimes
        }
        inSequence {
            (graphDB.readExternalProps(_: Request, _: List[String])).expects(*, List("hierarchy")).returns(Future(getRootExternalPropsResponseBL))
            (graphDB.readExternalProps(_: Request, _: List[String])).expects(*, List("hierarchy")).returns(Future(getNewRootExternalPropsResponseBL)).anyNumberOfTimes
        }
        val request = getQuestionSetCopyRequest()
        request.putAll(mapAsJavaMap(Map("identifier" -> "do_1234", "mode" -> "", "copyType" -> "deep")))
        request.setOperation("copyQuestionSet")
        val response = callActor(request, Props(new QuestionSetActor()))
        assert("successful".equals(response.getParams.getStatus))
    }
```

And the file with object creation methods is: https://github.com/project-sunbird/knowledge-platform/blob/f05328de25acba194e6210dda606e333ba12f6c7/assessment-api/assessment-actors/src/test/scala/org/sunbird/actors/copyTrait.scala 

Still not clear on how I can intercept the response and then get the new identifier to pass it to the object creation functions. 

Prototype of the getRootNodeWithBL is:
```
def getRootNodeWithBL(rootId: String, sectionId: String, addBranchingLogic: Boolean, withChildren: Boolean): Node = {
```

**Reply by Krishnaj20** [reply]
@aks30 & @kiranharidas187 - Can one of you can help @joffinjoy?

@alok-os

**Reply by aks30** [reply]
Sorry @Krishnaj20 but neither I or Kiran have an understanding of Scala and JAVA to help here.
